### PR TITLE
Duration

### PR DESCRIPTION
**Pull request by gabrielelanaro**
Summary:

We would like to introduce a new type called `Duration` to express time deltas not anchored to a specific date. Another way to interpret a duration is something that could be attached to a date to form an interval.

Date + Duration = interval

Things I don't like:
1. Format: the format for Time and Interval is some string that specifies date, hour and some other details, what would be a good format for duration? I made up a string format that is a bit different from `Time`, but it may make sense to put some more thoughts into it.

2. one night, ein ubernachtung is a duration of ~12-24 hours, with the caveat that those should cover a whole night. A simple timedelta data structure won't provide this information. We could however assumed that this irrelevant as those caveats can be quite complicated. Another example could be "2 hours in the morning". To account for this one would need to add a modifier: Duration(12 hours / night), although this seems very domain specific and will implement only if truly necessary.

Some insight into the problem is in here: https://kentonl.com/pub/ladz-acl.2014.pdf
Basically the time expressions could be represented using a sort of grammar that encodes those various details.

### CODE REVIEW

File: ctparse/time/rules.py [hunk 1]
```diff
@@ -783,14 +784,101 @@ def _adjust_h(t: Time) -> Optional[int]:
...
+    (21, r"twentyone|einunzwanzig"),
+    (22, r"twentytwo|zweiunzwanzig"),
+    (23, r"twentythree|dreiunzwanzig"),
+    (24, r"twentyfour|vierunzwanzig"),
+    (25, r"twentyfive|fünfunzwanzig"),
+    (26, r"twentysix|sechsunzwanzig"),
+    (27, r"twentyseven|siebenunzwanzig"),
+    (28, r"twentyeight|achtunzwanzig"),
+    (29, r"twentynine|neunundzwanzig"),
+    (30, r"thirty|dreißig"),
```
Comment (sebastianmika) [inline ctparse/time/rules.py:818]: And `dreissig` etc.

File: ctparse/time/rules.py [hunk 2]
```diff
@@ -783,14 +784,101 @@ def _adjust_h(t: Time) -> Optional[int]:
...
+    (31, r"thirtyone|einundreißig"),
+)
+_rule_named_number = "|".join(
+    r"(?P<n_{}>{}\b)".format(n, expr) for n, expr in _named_number
+)
+_rule_named_number = r"({})\s*".format(_rule_named_number)
+
+_durations = {
+    "night": r"n[aä]chte?|nights?",
```
Comment (sebastianmika) [inline ctparse/time/rules.py]: I would prefer plural (i.e. `nights`, `days`, etc.)

File: ctparse/types.py [hunk 3]
```diff
@@ -488,31 +488,16 @@ def end(self) -> Optional[Time]:
...
-        years: Optional[int] = None,
-        months: Optional[int] = None,
-        days: Optional[int] = None,
-        hours: Optional[int] = None,
-        minutes: Optional[int] = None,
-    ):
+    def __init__(self, value: int, unit: str):
+        """Create a Duration using value and unit.
+
+        Typical values for unit are:
```
Comment (sebastianmika) [inline ctparse/types.py:494]: Maybe type the unit as an enum?

**Comment by gabrielelanaro** [comment]
@sebastianmika please take another look, I took a stab at making a bunch of rules that make duration. I haven't tackled the TOD part because it require some careful design (for example dealing with veryearlyevening, first, last, and wether a Time with TOD is just an interval).

File: ctparse/time/rules.py [hunk 4]
```diff
@@ -779,3 +782,184 @@ def _adjust_h(t: Time) -> Optional[int]:
...
+    (12, r"twelve|zwölf"),
+    (13, r"thirteen|dreizehn"),
+    (14, r"fourteen|vierzehn"),
+    (15, r"fifteen|fünfzehn"),
+    (16, r"sixteen|sechszehn"),
+    (17, r"seventeen|siebzehn"),
+    (18, r"eighteen|achtzehn"),
+    (19, r"nineteen|neunzehn"),
+    (20, r"twenty|zwanzig"),
+    (21, r"twentyone|einunzwanzig"),
```
Comment (sebastianmika) [inline ctparse/time/rules.py:809]: `einunDzwanzig` - but probably a common type - ff. the same

File: ctparse/time/rules.py [hunk 5]
```diff
@@ -779,3 +782,184 @@ def _adjust_h(t: Time) -> Optional[int]:
...
+
+
+@rule(dimension(Time), r"f[üo]r", dimension(Duration))
+def ruleTimeDuration(
+    ts: datetime, t: Time, _: RegexMatch, dur: Duration
+) -> Optional[Interval]:
+    # Examples:
+    # on the 27th for one day
+    # heute eine Übernachtung
+    if t.hasDate:
```
Comment (sebastianmika) [inline ctparse/time/rules.py:925]: Then use `predicate("hasDate")` in the rule instead of `dimension(Time)`?

File: ctparse/time/rules.py [hunk 6]
```diff
@@ -779,3 +782,184 @@ def _adjust_h(t: Time) -> Optional[int]:
...
+        # To make an interval we should at least have a date
+        if (
+            dur.unit
+            in (
+                DurationUnit.DAYS,
+                DurationUnit.NIGHTS,
+                DurationUnit.WEEKS,
+                DurationUnit.MONTHS,
+            )
+            and t.hasDate
```
Comment (sebastianmika) [inline ctparse/time/rules.py:935]: If has already a date - or do you mean `isDate`?

File: ctparse/time/rules.py [hunk 7]
```diff
@@ -779,3 +782,184 @@ def _adjust_h(t: Time) -> Optional[int]:
...
+) -> Optional[Interval]:
+    # Examples:
+    # on the 27th for one day
+    # heute eine Übernachtung
+    if t.hasDate:
+        # To make an interval we should at least have a date
+        if (
+            dur.unit
+            in (
+                DurationUnit.DAYS,
```
Comment (sebastianmika) [inline ctparse/time/rules.py:930]: Maybe we should/could add some predicates on the `Duration` as well. That's what I thought on the `ruleDurationInterval` above. I haven't thought long about this, but they seem to mostly make sense for `predicate("isDateInterval") predicate("isDayNightDuration")` sequences OR `predicate("isTime") predicate("isHourOrMinuteDuration")` (clearly these are fake predicates that would need to be defined). Using predicates is a bit more expensive, but also reduces the number of branches. Not sure though whether this is super important with Durations as they are less common.

File: datasets/timeparse_corpus.json [hunk 8]
```diff
@@ -32,7 +32,7 @@
  {
   "text": "one night",
   "ref_time": "2017-10-03T14:30:35",
-  "gold_parse": "Interval[]{2017-10-28 X:X (X/X) - 2017-10-29 X:X (X/X)}",
+  "gold_parse": "Duration[]{1 nights}",
```
Comment (sebastianmika) [inline datasets/timeparse_corpus.json:35]: Yes!

File: ctparse/time/rules.py (hunk 5 above)
Comment (gabrielelanaro) [reply ctparse/time/rules.py:925]: quite neat

File: ctparse/time/rules.py (hunk 6 above)
Comment (gabrielelanaro) [reply ctparse/time/rules.py:935]: it was a leftover when I moved the if top-level
